// rettiwt-api/dist/collections/Extractors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.extractors = void 0;
var Data_1 = require("../enums/Data");
var CursoredData_1 = require("../models/data/CursoredData");
var Tweet_1 = require("../models/data/Tweet");
var User_1 = require("../models/data/User");
/**
 * Collection of data extractors for each resource.
 *
 * @internal
 */
exports.extractors = {
    /* eslint-disable @typescript-eslint/naming-convention */
    LIST_TWEETS: function (response) {
        return new CursoredData_1.CursoredData(response, Data_1.EBaseType.TWEET);
    },
    MEDIA_UPLOAD_APPEND: function () { return undefined; },
    MEDIA_UPLOAD_FINALIZE: function () { return undefined; },
    MEDIA_UPLOAD_INITIALIZE: function (response) { var _a; return (_a = response.media_id_string) !== null && _a !== void 0 ? _a : undefined; },
    TWEET_DETAILS: function (response, id) { return Tweet_1.Tweet.single(response, id); },
    TWEET_DETAILS_ALT: function (response, id) { return Tweet_1.Tweet.single(response, id); },
    TWEET_LIKE: function (response) { var _a; return (((_a = response === null || response === void 0 ? void 0 : response.data) === null || _a === void 0 ? void 0 : _a.favorite_tweet) ? true : false); },
    TWEET_POST: function (response) { var _a, _b, _c, _d, _e; return (_e = (_d = (_c = (_b = (_a = response === null || response === void 0 ? void 0 : response.data) === null || _a === void 0 ? void 0 : _a.create_tweet) === null || _b === void 0 ? void 0 : _b.tweet_results) === null || _c === void 0 ? void 0 : _c.result) === null || _d === void 0 ? void 0 : _d.rest_id) !== null && _e !== void 0 ? _e : undefined; },
    TWEET_RETWEET: function (response) { var _a; return (((_a = response === null || response === void 0 ? void 0 : response.data) === null || _a === void 0 ? void 0 : _a.create_retweet) ? true : false); },
    TWEET_RETWEETERS: function (response) {
        return new CursoredData_1.CursoredData(response, Data_1.EBaseType.USER);
    },
    TWEET_SCHEDULE: function (response) { var _a, _b, _c; return (_c = (_b = (_a = response === null || response === void 0 ? void 0 : response.data) === null || _a === void 0 ? void 0 : _a.tweet) === null || _b === void 0 ? void 0 : _b.rest_id) !== null && _c !== void 0 ? _c : undefined; },
    TWEET_SEARCH: function (response) {
        return new CursoredData_1.CursoredData(response, Data_1.EBaseType.TWEET);
    },
    TWEET_UNLIKE: function (response) { var _a; return (((_a = response === null || response === void 0 ? void 0 : response.data) === null || _a === void 0 ? void 0 : _a.unfavorite_tweet) ? true : false); },
    TWEET_UNPOST: function (response) { var _a; return (((_a = response === null || response === void 0 ? void 0 : response.data) === null || _a === void 0 ? void 0 : _a.delete_tweet) ? true : false); },
    TWEET_UNRETWEET: function (response) { var _a, _b, _c; return ((_c = (_b = (_a = response === null || response === void 0 ? void 0 : response.data) === null || _a === void 0 ? void 0 : _a.unretweet) === null || _b === void 0 ? void 0 : _b.source_tweet_results) === null || _c === void 0 ? void 0 : _c.result) ? true : false; },
    TWEET_UNSCHEDULE: function (response) { var _a; return ((_a = response === null || response === void 0 ? void 0 : response.data) === null || _a === void 0 ? void 0 : _a.scheduledtweet_delete) == 'Done'; },
    USER_DETAILS_BY_USERNAME: function (response) { return User_1.User.single(response); },
    USER_DETAILS_BY_ID: function (response) { return User_1.User.single(response); },
    USER_FEED_FOLLOWED: function (response) {
        return new CursoredData_1.CursoredData(response, Data_1.EBaseType.TWEET);
    },
    USER_FEED_RECOMMENDED: function (response) {
        return new CursoredData_1.CursoredData(response, Data_1.EBaseType.TWEET);
    },
    USER_FOLLOW: function (response) { return ((response === null || response === void 0 ? void 0 : response.id) ? true : false); },
    USER_FOLLOWING: function (response) {
        return new CursoredData_1.CursoredData(response, Data_1.EBaseType.USER);
    },
    USER_FOLLOWERS: function (response) {
        return new CursoredData_1.CursoredData(response, Data_1.EBaseType.USER);
    },
    USER_HIGHLIGHTS: function (response) {
        return new CursoredData_1.CursoredData(response, Data_1.EBaseType.TWEET);
    },
    USER_LIKES: function (response) {
        return new CursoredData_1.CursoredData(response, Data_1.EBaseType.TWEET);
    },
    USER_MEDIA: function (response) {
        return new CursoredData_1.CursoredData(response, Data_1.EBaseType.TWEET);
    },
    USER_NOTIFICATIONS: function (response) {
        return new CursoredData_1.CursoredData(response, Data_1.EBaseType.NOTIFICATION);
    },
    USER_SUBSCRIPTIONS: function (response) {
        return new CursoredData_1.CursoredData(response, Data_1.EBaseType.USER);
    },
    USER_TIMELINE: function (response) {
        return new CursoredData_1.CursoredData(response, Data_1.EBaseType.TWEET);
    },
    USER_TIMELINE_AND_REPLIES: function (response) {
        return new CursoredData_1.CursoredData(response, Data_1.EBaseType.TWEET);
    },
    USER_UNFOLLOW: function (response) { return ((response === null || response === void 0 ? void 0 : response.id) ? true : false); },
    /* eslint-enable @typescript-eslint/naming-convention */
};
