// rettiwt-core/dist/models/params/Variables.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReplyVariable = exports.MediaEntityVariable = exports.MediaVariable = void 0;
/**
 * Media to be sent as payload.
 *
 * @internal
 */
var MediaVariable = /** @class */ (function () {
    /* eslint-enable @typescript-eslint/naming-convention */
    /**
     * @param media - The list of NewTweetMedia objects specifying the media items to be sent in the Tweet.
     */
    function MediaVariable(media) {
        this.media_entities = media.map(function (item) { return new MediaEntityVariable(item); });
        this.possibly_sensitive = false;
    }
    return MediaVariable;
}());
exports.MediaVariable = MediaVariable;
/**
 * Each media item in the media payload.
 *
 * @internal
 */
var MediaEntityVariable = /** @class */ (function () {
    /* eslint-enable @typescript-eslint/naming-convention */
    /**
     * @param media - The NewTweetMedia object specifying the details of the media item to be included in the payload.
     */
    function MediaEntityVariable(media) {
        var _a;
        this.media_id = media.id;
        this.tagged_users = (_a = media.tags) !== null && _a !== void 0 ? _a : [];
    }
    return MediaEntityVariable;
}());
exports.MediaEntityVariable = MediaEntityVariable;
/**
 * Reply specific details to be sent in payload.
 *
 * @internal
 */
var ReplyVariable = /** @class */ (function () {
    /* eslint-enable @typescript-eslint/naming-convention */
    /**
     * @param replyTo - The id of the Tweet to which this Tweet is a reply.
     */
    function ReplyVariable(replyTo) {
        this.in_reply_to_tweet_id = replyTo;
        this.exclude_reply_user_ids = [];
    }
    return ReplyVariable;
}());
exports.ReplyVariable = ReplyVariable;
