// rettiwt-auth/dist/models/payloads/subtask_inputs/UserIdentifier.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UserIdentifierInput = void 0;
/**
 * The input data to be sent for identifying the user using email, phone, etc.
 */
var UserIdentifierInput = /** @class */ (function () {
    /* eslint-enable @typescript-eslint/naming-convention */
    /**
     * @param userId - The id to be used for identifying the user.
     */
    function UserIdentifierInput(userId) {
        this.link = 'next_link';
        this.setting_responses = [new SettingResponse(userId)];
    }
    return UserIdentifierInput;
}());
exports.UserIdentifierInput = UserIdentifierInput;
/**
 * The response received from the user for identification purpose.
 */
var SettingResponse = /** @class */ (function () {
    /* eslint-enable @typescript-eslint/naming-convention */
    /**
     * @param responseText - The text entered by the user.
     */
    function SettingResponse(responseText) {
        /* eslint-disable @typescript-eslint/naming-convention */
        this.key = 'user_identifier';
        this.response_data = new ResponseData(responseText);
    }
    return SettingResponse;
}());
/**
 * The raw response data in case of identification using email.
 */
var ResponseData = /** @class */ (function () {
    /* eslint-enable */
    /**
     * @param email - The email id of the user.
     */
    function ResponseData(email) {
        this.text_data = new TextData(email);
    }
    return ResponseData;
}());
/**
 * The raw, elemental text data.
 */
var TextData = /** @class */ (function () {
    /**
     * @param text - The text data entered.
     */
    function TextData(text) {
        this.result = text;
    }
    return TextData;
}());
