// rettiwt-api/dist/enums/Api.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EErrorCodes = exports.EApiErrors = void 0;
/**
 * The different types of api error messages.
 *
 * @public
 */
var EApiErrors;
(function (EApiErrors) {
    EApiErrors["COULD_NOT_AUTHENTICATE"] = "Failed to authenticate";
    EApiErrors["RESOURCE_NOT_FOUND"] = "Requested resource not found";
    EApiErrors["MISSING_PARAMETER"] = "Missing named parameter";
    EApiErrors["USER_NOT_FOUND"] = "Requested user not found";
    EApiErrors["USER_SUSPENDED"] = "Requested user has been suspended";
    EApiErrors["ACCOUNT_SUSPENDED"] = "Account has been suspended";
    EApiErrors["RATE_LIMIT_EXCEEDED"] = "Rate limit exceeded";
    EApiErrors["INTERNAL_ERROR"] = "Internal server error";
    EApiErrors["TIME_ERROR"] = "Mismatched data/time with server";
    EApiErrors["ALREADY_FAVORITED"] = "Tweet already favorited";
    EApiErrors["STATUS_NOT_FOUND"] = "Requested tweeet not found";
    EApiErrors["NOT_AUTHORIZED"] = "Not authorized to view tweet";
    EApiErrors["DAILY_STATUS_LIMIT_EXCEEDED"] = "Exceeded daily tweet update limit";
    EApiErrors["TWEET_LENGTH_EXCEEDED"] = "Exceeded tweet text maximum length";
    EApiErrors["DUPLICATE_STATUS"] = "Tweet already posted";
    EApiErrors["BAD_AUTHENTICATION"] = "Invalid authentication data";
    EApiErrors["RESOURCE_NOT_ALLOWED"] = "Not authorized to access requested resource";
    EApiErrors["AUTOMATED_REQUEST_ERROR"] = "Automated request detected";
    EApiErrors["ACCOUNT_LOCKED"] = "Account has been locked";
    EApiErrors["ALREADY_RETWEETED"] = "Tweet already retweeted";
    EApiErrors["TWEET_NOT_FOUND"] = "Requested tweet not found";
    EApiErrors["TWEET_VIOLATED_RULES"] = "Requestd tweet has been removed for rules violation";
    EApiErrors["DISABLED_TWEET_ACTIONS"] = "Reqeusted action disabled on the tweet";
})(EApiErrors || (exports.EApiErrors = EApiErrors = {}));
/**
 * The different error codes.
 *
 * @public
 */
var EErrorCodes;
(function (EErrorCodes) {
    EErrorCodes[EErrorCodes["COULD_NOT_AUTHENTICATE"] = 32] = "COULD_NOT_AUTHENTICATE";
    EErrorCodes[EErrorCodes["RESOURCE_NOT_FOUND"] = 34] = "RESOURCE_NOT_FOUND";
    EErrorCodes[EErrorCodes["MISSING_PARAMETER"] = 38] = "MISSING_PARAMETER";
    EErrorCodes[EErrorCodes["USER_NOT_FOUND"] = 50] = "USER_NOT_FOUND";
    EErrorCodes[EErrorCodes["USER_SUSPENDED"] = 63] = "USER_SUSPENDED";
    EErrorCodes[EErrorCodes["ACCOUNT_SUSPENDED"] = 64] = "ACCOUNT_SUSPENDED";
    EErrorCodes[EErrorCodes["RATE_LIMIT_EXCEEDED"] = 88] = "RATE_LIMIT_EXCEEDED";
    EErrorCodes[EErrorCodes["INTERNAL_ERROR"] = 131] = "INTERNAL_ERROR";
    EErrorCodes[EErrorCodes["TIME_ERROR"] = 135] = "TIME_ERROR";
    EErrorCodes[EErrorCodes["ALREADY_FAVORITED"] = 139] = "ALREADY_FAVORITED";
    EErrorCodes[EErrorCodes["STATUS_NOT_FOUND"] = 144] = "STATUS_NOT_FOUND";
    EErrorCodes[EErrorCodes["NOT_AUTHORIZED"] = 179] = "NOT_AUTHORIZED";
    EErrorCodes[EErrorCodes["DAILY_STATUS_LIMIT_EXCEEDED"] = 185] = "DAILY_STATUS_LIMIT_EXCEEDED";
    EErrorCodes[EErrorCodes["TWEET_LENGTH_EXCEEDED"] = 186] = "TWEET_LENGTH_EXCEEDED";
    EErrorCodes[EErrorCodes["DUPLICATE_STATUS"] = 187] = "DUPLICATE_STATUS";
    EErrorCodes[EErrorCodes["BAD_AUTHENTICATION"] = 215] = "BAD_AUTHENTICATION";
    EErrorCodes[EErrorCodes["RESOURCE_NOT_ALLOWED"] = 220] = "RESOURCE_NOT_ALLOWED";
    EErrorCodes[EErrorCodes["AUTOMATED_REQUEST_ERROR"] = 226] = "AUTOMATED_REQUEST_ERROR";
    EErrorCodes[EErrorCodes["ACCOUNT_LOCKED"] = 326] = "ACCOUNT_LOCKED";
    EErrorCodes[EErrorCodes["ALREADY_RETWEETED"] = 327] = "ALREADY_RETWEETED";
    EErrorCodes[EErrorCodes["TWEET_NOT_FOUND"] = 421] = "TWEET_NOT_FOUND";
    EErrorCodes[EErrorCodes["TWEET_VIOLATED_RULES"] = 422] = "TWEET_VIOLATED_RULES";
    EErrorCodes[EErrorCodes["DISABLED_TWEET_ACTIONS"] = 425] = "DISABLED_TWEET_ACTIONS";
})(EErrorCodes || (exports.EErrorCodes = EErrorCodes = {}));
