// rettiwt-api/dist/enums/Resource.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EResourceType = void 0;
/**
 * The different types of resources that can be interacted with.
 *
 * @public
 */
var EResourceType;
(function (EResourceType) {
    // LIST
    EResourceType["LIST_TWEETS"] = "LIST_TWEETS";
    // MEDIA
    EResourceType["MEDIA_UPLOAD_APPEND"] = "MEDIA_UPLOAD_APPEND";
    EResourceType["MEDIA_UPLOAD_FINALIZE"] = "MEDIA_UPLOAD_FINALIZE";
    EResourceType["MEDIA_UPLOAD_INITIALIZE"] = "MEDIA_UPLOAD_INITIALIZE";
    // TWEET
    EResourceType["TWEET_DETAILS"] = "TWEET_DETAILS";
    EResourceType["TWEET_DETAILS_ALT"] = "TWEET_DETAILS_ALT";
    EResourceType["TWEET_LIKE"] = "TWEET_LIKE";
    EResourceType["TWEET_POST"] = "TWEET_POST";
    EResourceType["TWEET_RETWEET"] = "TWEET_RETWEET";
    EResourceType["TWEET_RETWEETERS"] = "TWEET_RETWEETERS";
    EResourceType["TWEET_SCHEDULE"] = "TWEET_SCHEDULE";
    EResourceType["TWEET_SEARCH"] = "TWEET_SEARCH";
    EResourceType["TWEET_UNLIKE"] = "TWEET_UNLIKE";
    EResourceType["TWEET_UNPOST"] = "TWEET_UNPOST";
    EResourceType["TWEET_UNRETWEET"] = "TWEET_UNRETWEET";
    EResourceType["TWEET_UNSCHEDULE"] = "TWEET_UNSCHEDULE";
    // USER
    EResourceType["USER_DETAILS_BY_USERNAME"] = "USER_DETAILS_BY_USERNAME";
    EResourceType["USER_DETAILS_BY_ID"] = "USER_DETAILS_BY_ID";
    EResourceType["USER_FEED_FOLLOWED"] = "USER_FEED_FOLLOWED";
    EResourceType["USER_FEED_RECOMMENDED"] = "USER_FEED_RECOMMENDED";
    EResourceType["USER_FOLLOW"] = "USER_FOLLOW";
    EResourceType["USER_FOLLOWING"] = "USER_FOLLOWING";
    EResourceType["USER_FOLLOWERS"] = "USER_FOLLOWERS";
    EResourceType["USER_HIGHLIGHTS"] = "USER_HIGHLIGHTS";
    EResourceType["USER_LIKES"] = "USER_LIKES";
    EResourceType["USER_MEDIA"] = "USER_MEDIA";
    EResourceType["USER_SUBSCRIPTIONS"] = "USER_SUBSCRIPTIONS";
    EResourceType["USER_TIMELINE"] = "USER_TIMELINE";
    EResourceType["USER_TIMELINE_AND_REPLIES"] = "USER_TIMELINE_AND_REPLIES";
    EResourceType["USER_UNFOLLOW"] = "USER_UNFOLLOW";
})(EResourceType || (exports.EResourceType = EResourceType = {}));
