// rettiwt-api/dist/models/errors/DataValidationError.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ValidationErrorDetails = exports.DataValidationError = void 0;
/**
 * Represents and error when any fields of a validation-enabled class fails to validate.
 *
 * @public
 */
var DataValidationError = /** @class */ (function () {
    /**
     * @param data - The error details, as a list of type {@link ValidationError}
     */
    function DataValidationError(errorDetails) {
        this.name = 'VALIDATION_ERROR';
        this.message = 'One or more validation error(s) occured, check details field.';
        this.details = errorDetails.map(function (error) { return new ValidationErrorDetails(error); });
    }
    return DataValidationError;
}());
exports.DataValidationError = DataValidationError;
/**
 * Represents the validation error details of a single field.
 *
 * @public
 */
var ValidationErrorDetails = /** @class */ (function () {
    function ValidationErrorDetails(details) {
        this.field = details.property;
        this.constraints = Object.values(details.constraints);
    }
    return ValidationErrorDetails;
}());
exports.ValidationErrorDetails = ValidationErrorDetails;
